src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +526,437 @@ pub fn process_participation_record_updates<
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for source-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_source_attestations =
+        get_matching_source_attestations(state, previous_epoch, context)?;
+    get_attestation_component_deltas(state, matching_source_attestations, context)
+}
+
+pub fn get_target_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for target-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_target_attestations =
+        get_matching_target_attestations(state, previous_epoch, context)?
+            .cloned()
+            .collect::<Vec<_>>();
+    get_attestation_component_deltas(state, matching_target_attestations.as_slice(), context)
+}
+
+pub fn get_head_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for head-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_head_attestations =
+        get_matching_head_attestations(state, previous_epoch, context)?
+            .cloned()
+            .collect::<Vec<_>>();
+    get_attestation_component_deltas(state, matching_head_attestations.as_slice(), context)
```

Comment:
random nitpick, but as far as i understand, it is a bit more idiomatic to use `&t` for `t: Vec<T>` rather than call the method `as_slice`

Reply:
Agreed, idk why I used that hah